src/rayon_impls.rs

```diff
@@ -0,0 +1,119 @@
+use crate::HashMap;
+use rayon::iter::{FromParallelIterator, IntoParallelIterator, ParallelExtend, ParallelIterator};
+use std::hash::{BuildHasher, Hash};
+
+impl<K, V, S> ParallelExtend<(K, V)> for HashMap<K, V, S>
+where
+    K: Clone + Hash + Ord + Send + Sync + 'static,
+    V: Send + Sync + 'static,
+    S: BuildHasher + Sync,
+{
+    // This is of limited use due to the `&mut self` parameter. See `par_extend_sync`
+    fn par_extend<I>(&mut self, par_iter: I)
+    where
+        I: IntoParallelIterator<Item = (K, V)>,
+    {
+        self.par_extend_sync(par_iter);
+    }
+}
+
+impl<K, V, S> HashMap<K, V, S>
+where
+    K: Clone + Hash + Ord + Send + Sync + 'static,
+    V: Send + Sync + 'static,
+    S: BuildHasher + Sync,
+{
+    // FIXME: Terrible name, just didn't want to shadow the rayon name
+    fn par_extend_sync<I>(&self, par_iter: I)
+    where
+        I: IntoParallelIterator<Item = (K, V)>,
+    {
+        par_iter.into_par_iter().for_each(|(k, v)| {
+            // Unfortunate that we need to create a guard for each insert operation
+            // Ideally we'd create a guard for each `rayon` worker thread instead
+            // Perhaps this could be done with a thread local?
+            let guard = self.guard();
```

Comment:
I think [`for_each_init`](https://docs.rs/rayon/1.3.0/rayon/iter/trait.ParallelIterator.html#method.for_each_init) would be perfect for creating that guard. It's per job, not exactly per thread, which is different depending on how adaptive splitting and work stealing plays out, but close enough most of the time.